File: BubblerAndroid/app/src/main/java/com/bubbler/android/data/model/StaffReport.kt
package com.bubbler.android.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.Instant

@Serializable
enum class StaffReportStatus {
    @SerialName("open")
    OPEN,

    @SerialName("in_review")
    IN_REVIEW,

    @SerialName("resolved")
    RESOLVED,

    @SerialName("dismissed")
    DISMISSED;

    val title: String
        get() = when (this) {
            OPEN -> "Open"
            IN_REVIEW -> "In review"
            RESOLVED -> "Resolved"
            DISMISSED -> "Dismissed"
        }

    val apiValue: String
        get() = when (this) {
            OPEN -> "open"
            IN_REVIEW -> "in_review"
            RESOLVED -> "resolved"
            DISMISSED -> "dismissed"
        }
}

/** Staff review ticket (`GET /admin/reports`). */
@Serializable
data class StaffReport(
    val id: String,
    @SerialName("reporter_id")
    val reporterId: Int? = null,
    @SerialName("post_id")
    val postId: String? = null,
    @SerialName("reported_user_id")
    val reportedUserId: Int? = null,
    val reason: String,
    val details: String? = null,
    val status: StaffReportStatus,
    @SerialName("content_snapshot")
    val contentSnapshot: String,
    @SerialName("topic_snapshot")
    val topicSnapshot: String? = null,
    @SerialName("author_username_snapshot")
    val authorUsernameSnapshot: String? = null,
    @SerialName("created_at")
    @Serializable(with = InstantIsoSerializer::class)
    val createdAt: Instant,
) {
    val reasonTitle: String
        get() = when (reason) {
            "illegal_content" -> "Illegal content"
            "severe_violence" -> "Severe violence"
            "non_consensual_sexual_content" -> "Non-consensual sexual content"
            "harassment" -> "Harassment or threats"
            "spam" -> "Spam"
            "other" -> "Other"
            else -> reason.replace('_', ' ').replaceFirstChar { it.uppercase() }
        }
}

@Serializable
data class StaffReportStatusUpdateBody(
    val status: String,
)

File: BubblerApp/BubblerApp/Models/StaffReport.swift
import Foundation

enum StaffReportStatus: String, CaseIterable, Codable, Identifiable {
    case open
    case inReview = "in_review"
    case resolved
    case dismissed

    var id: String { rawValue }

    var title: String {
        switch self {
        case .open: "Open"
        case .inReview: "In review"
        case .resolved: "Resolved"
        case .dismissed: "Dismissed"
        }
    }
}

struct StaffReport: Codable, Identifiable, Equatable {
    let id: String
    let reporterId: Int?
    let postId: String?
    let reportedUserId: Int?
    let reason: String
    let details: String?
    let status: StaffReportStatus
    let contentSnapshot: String
    let topicSnapshot: String?
    let authorUsernameSnapshot: String?
    let createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id
        case reporterId = "reporter_id"
        case postId = "post_id"
        case reportedUserId = "reported_user_id"
        case reason
        case details
        case status
        case contentSnapshot = "content_snapshot"
        case topicSnapshot = "topic_snapshot"
        case authorUsernameSnapshot = "author_username_snapshot"
        case createdAt = "created_at"
    }

    var reasonTitle: String {
        switch reason {
        case "illegal_content": "Illegal content"
        case "severe_violence": "Severe violence"
        case "non_consensual_sexual_content": "Non-consensual sexual content"
        case "harassment": "Harassment or threats"
        case "spam": "Spam"
        case "other": "Other"
        default: reason.replacingOccurrences(of: "_", with: " ").capitalized
        }
    }
}

